opm/simulators/wells/GasLiftStage2.cpp

```diff
@@ -1075,13 +963,35 @@ checkGasTarget()
     }
     return false;
 }
+bool
+GasLiftStage2::SurplusState::
+checkLiquidTarget(double eco_grad)
+{
+    if (this->group.has_control(Group::ProductionCMode::LRAT)) {
+        auto liquid_rate = this->oil_rate + this->water_rate;
+        // the eco_grad is subtracted from the rate to make sure the
+        // group still can produce its target
+        if (this->liquid_target < (liquid_rate - eco_grad) ) {
+            if (this->parent.debug) {
+                const std::string msg = fmt::format("group: {} : "
+                    "liquid rate {} is greater than liquid target {}", this->group.name(),
+                    liquid_rate, this->liquid_target);
+                this->parent.displayDebugMessage_(msg);
+            }
+            return true;
+        }
+    }
+    return false;
+}
 
 bool
 GasLiftStage2::SurplusState::
-checkOilTarget()
+checkOilTarget(double eco_grad)
 {
     if (this->group.has_control(Group::ProductionCMode::ORAT)) {
-        if (this->oil_target < this->oil_rate  ) {
+        // the eco_grad is subtracted from the rate to make sure the
+        // group still can produce its target
+        if (this->oil_target < (this->oil_rate - eco_grad)  ) {
```

Comment:
Thanks for the update! One final question: Should `eco_grad` be multiplied by the alq increment, to convert it into an oil rate? Something like: `auto delta_oil = eco_grad * this->parent.glo_.gaslift_increment()` and then the if statement: `if (this->oil_target < (this->oil_rate - delta_oil)  ) {`

Reply:
Thanks for the comment. I though the eco_grad already was the delta_rate, but your're of-course right. Is it possible to get the delta_water etc from the gradient?

Reply:
If you look at lines 1018-1021 in `updateRates()`, we compute `delta_oil`, `delta_water`, ... Maybe we could save these values or recompute them here? Instead of using the `eco_grad` to compute them.

Reply:
I will try to split the method in two. First to compute the deltas = computeDeltas(..) , then later apply them in updateRates( deltas ) if remove = true.